waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -0,0 +1,594 @@
+   (let [pod-name (get-in pod [:metadata :name])
+         instance-suffix (subs pod-name (- (count pod-name) pod-unique-suffix-length))
+         service-id (get-in pod [:metadata :annotations :waiter-service-id])]
+     (str service-id \. instance-suffix \- restart-count))))
+
+(defn- killed-by-k8s?
+  "Determine whether a pod was killed (restarted) by its corresponding Kubernetes liveness checks."
+  [pod-terminated-info]
+  ;; TODO (#351) - Look at events for messages about liveness probe failures.
+  ;; Currently, we assume any SIGKILL (137) with the default "Error" reason was a livenessProbe kill.
+  (and (= 137 (:exitCode pod-terminated-info))
+       (= "Error" (:reason pod-terminated-info))))
+
+(defn- track-failed-instances!
+  "Update this KubernetesScheduler's service-id->failed-instances-transient-store
+   when a new pod failure is listed in the given pod's lastState container status.
+   Note that unique instance-ids are deterministically generated each time the pod is restarted
+   by passing the pod's restartCount value to the pod->instance-id function."
+  [{:keys [service-id] :as live-instance} {:keys [service-id->failed-instances-transient-store]} pod]
+  (when-let [newest-failure (get-in pod [:status :containerStatuses 0 :lastState :terminated])]
+    (let [failure-flags (if (= "OOMKilled" (:reason newest-failure)) #{:memory-limit-exceeded} #{})
+          newest-failure-start-time (-> newest-failure :startedAt timestamp-str->datetime)
+          restart-count (get-in pod [:status :containerStatuses 0 :restartCount])
+          newest-failure-id (pod->instance-id pod (dec restart-count))
+          failures (-> service-id->failed-instances-transient-store deref (get service-id))]
+      (when-not (contains? failures newest-failure-id)
+        (let [newest-failure-instance (merge live-instance
+                                             ;; To match the behavior of the marathon scheduler,
+                                             ;; we don't include the exit code in failed instances that were killed by k8s.
+                                             (when-not (killed-by-k8s? newest-failure)
+                                               {:exit-code (:exitCode newest-failure)})
+                                             {:flags failure-flags
+                                              :healthy? false
+                                              :id newest-failure-id
+                                              :started-at newest-failure-start-time})]
+          (swap! service-id->failed-instances-transient-store
+                 update-in [service-id] assoc newest-failure-id newest-failure-instance))))))
+
+(defn- pod->ServiceInstance
+  "Convert a Kubernetes Pod JSON response into a Waiter Service Instance record."
+  [pod]
+  (try
+    (let [port0 (get-in pod [:spec :containers 0 :ports 0 :containerPort])]
+      (scheduler/make-ServiceInstance
+        {:k8s-name (get-in pod [:metadata :labels :app])
```

Comment:
We should use namespace keywords for k8s specific entries, e.g.`k8s/name`, `k8s/namespace`, etc.

Reply:
Done.